gh-91172: Create a workflow for verifying bundled pip and setuptools

illia-v opened the pull request:
<!-- issue-number: [bpo-47016](https://bugs.python.org/issue47016) -->
https://bugs.python.org/issue47016
<!-- /issue-number -->

illia-v committed:
bpo-47016: Create a workflow for verifying bundled pip and setuptools

illia-v committed:
Revert "Corrupt the bundled pip wheel to test the new workflow"

This reverts commit e46f87d55ae60ebb45cfd0ba709614662fc48616.

illia-v committed:
Corrupt the bundled pip wheel to test the new workflow

illia-v committed:
Revert "Corrupt the bundled pip wheel to test the new workflow"

This reverts commit a57cd3ab63c186d607b91fed859c9c82087ac494.

illia-v committed:
Corrupt the bundled pip wheel to test the new workflow

illia-v committed:
Fix naming style of the new workflow

illia-v committed:
Allow manual triggering the new workflow

hugovk commented on .github/workflows/ensurepip_verify_bundled.yml:
```diff
@@ -0,0 +1,51 @@
+  verify:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v3
```

hugovk commented on .github/workflows/ensurepip_verify_bundled.yml:
```diff
@@ -0,0 +1,51 @@
+      - uses: actions/checkout@v2
+      - name: Compare checksums of bundled pip and setuptools to ones published on PyPI
+        run: |
+          package_names=("pip" "setuptools")
```
Perhaps this could be moved to a script (say, under `Tools/scripts`), so it can also be run locally?

illia-v replied:
Good idea, but it looks like all scripts in `Tools/scripts` are written in Python.
This is a Bash script, and it may not work on some platforms.

Do you know how compatible with different platforms scripts located in `Tools/scripts` should be?

hugovk replied:
Hmm, or perhaps the `Misc` directory? The [devguide](https://devguide.python.org/?highlight=misc#additional-resources) says:

> Various tools with configuration files as found in the [Misc directory](https://github.com/python/cpython/tree/main/Misc)

I don't know what compatibility is needed, but as this script is primarily intended for Ubuntu on the CI, I'd stick with that.

illia-v replied:
Done. [This is](https://github.com/illia-v/cpython/runs/5557015587?check_suite_focus=true) the latest run after I corrupted the pip wheel temporarily.

BTW, I added messages that GitHub has to show as annotations. But I could not see the annotations, maybe it does not show them for binary files…

hugovk replied:
They show up on the Summary page, for example clicking "🏠 Summary" at the top left: 

![image](https://user-images.githubusercontent.com/1324225/158424644-50c75f51-abc2-41ca-89d6-3549e23276f0.png)

 👍

illia-v committed:
Bump actions/checkout to v3

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

illia-v committed:
Create a separate script for verifying bundled wheels

hugovk commented on .github/workflows/verify_bundled_wheels.yml:
```diff
@@ -0,0 +1,18 @@
+name: Verify bundled pip and setuptools
```
Minor suggestion: I know the only other multi-word yml filename in https://github.com/python/cpython/tree/main/.github/workflows uses an underscore:

```console
$ ls .github/workflows
build.yml         build_msi.yml     doc.yml           posix-deps-apt.sh stale.yml
```

But I'd probably go for dashes here, like `posix-deps-apt.sh`.

illia-v replied:
Thank you for the review!

> But I'd probably go for dashes here, like `posix-deps-apt.sh`.

I renamed the file.

hugovk reviewed (approved):
Thank you! 🕊️🇺🇦✌️

illia-v committed:
Rename the workflow file

AA-Turner commented on Misc/verify-bundled-wheels.sh:
```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
```
Is there a reason this is written as a shell script? If (as @hugovk mentioned) it would be useful to run locally, rewriting in Python would make this more portable cross-platform (e.g. Windows).

Happy to help rewriting, just don't want to do the work if there's a requirement to have it be a shell script I'm missing!

A

illia-v replied:
> Is there a reason this is written as a shell script?

[Initially](https://github.com/python/cpython/pull/31885/commits/e4a8faf7b4e4aeb553e246d7e58d887a018dc341#diff-d7623692dd445a8dd8e63aaa1a557c50c8cd4e530aee2634865caa0b402a69a8) this script was embedded in the workflow file, making it a shell script seemed to be cleaner at the time.

pradyunsg replied:
I think this can be a Python script, but changing over to that should not be a blocking concern for the PR. :)

illia-v committed:
Merge branch 'main' into bpo-47016

AA-Turner committed:
Add verify-ensurepip-wheels.py

AA-Turner committed:
Update verify-bundled-wheels.yml

illia-v committed:
Make workflow permissions explicit

AA-Turner committed:
Add shebang and file mode permissions for unix users

AA-Turner committed:
git mv verify-ensurepip-wheels verify_ensurepip_wheels

AA-Turner committed:
git mv verify-bundled-wheels verify-ensurepip-wheels

AA-Turner committed:
Address review

AA-Turner committed:
Merge remote-tracking branch 'illia-v/bpo-47016' into bpo-47016

illia-v committed:
Merge pull request #1 from AA-Turner/bpo-47016

bpo-47016: Create a workflow for verifying bundled pip and setuptools

illia-v committed:
Delete the shell script

illia-v committed:
Mention Adam Turner in the news entry

illia-v committed:
Revert "Corrupt the bundled pip wheel to test the updated workflow"

This reverts commit 6786960784ae0de3962b94bf8b5c5a2f07c41c59.

illia-v committed:
Corrupt the bundled pip wheel to test the updated workflow

illia-v committed:
Revert "Corrupt the bundled pip wheel to test the updated workflow"

This reverts commit 6492602464fb56a9dc4ce510ab1405e3cf67f757.

illia-v committed:
Corrupt the bundled pip wheel to test the updated workflow

illia-v committed:
Refactor the script to fix the test

AA-Turner commented on Tools/scripts/verify_ensurepip_wheels.py:
```diff
@@ -0,0 +1,96 @@
+
+if __name__ == "__main__":
+    success = all([verify_wheel(package_name) for package_name in PACKAGE_NAMES])
+    sys.exit(0 if success else 1)
```
Don't construct an inner list, eliminate the `sys` import:
```suggestion
    success = all(verify_wheel(package_name) for package_name in PACKAGE_NAMES)
    if success:
        raise SystemExit(0)
    raise SystemExit(1)
```

illia-v replied:
The list comprehension was needed to verify both wheels even if the first verification fails. I've made this piece of code more explicit.

AA-Turner commented on Tools/scripts/verify_ensurepip_wheels.py:
```diff
@@ -0,0 +1,96 @@
+    # Compute the SHA 256 digest of the wheel on disk
+    actual_digest = hashlib.sha256(package_path.read_bytes()).hexdigest()
+
+    print(f"Expected digest: {expected_digest}")
+    print(f"Actual digest:   {actual_digest}")
```
```suggestion
    print_notice(package_path, f"Expected digest: {expected_digest}")
    print_notice(package_path, f"Actual digest:   {actual_digest}")
```

illia-v replied:
If I make the change, this will add redundant annotation bloat.

I think it is enough to state that a wheel has either been verified or not like in this two cases:
* https://github.com/illia-v/cpython/actions/runs/2455864579
* https://github.com/python/cpython/actions/runs/2455851364

AA-Turner replied:
Ahh, I didn't realise it was that prominent -- in that case makes sense to keep as current.

A

AA-Turner commented on Tools/scripts/verify_ensurepip_wheels.py:
```diff
@@ -0,0 +1,96 @@
+        print_error("", f"Could not find a {package_name} wheel on disk.")
+        return False
+
+    print(f"Verifying checksum for {package_path}.")
```
```suggestion
    print_notice(package_path, f"Verifying checksum for {package_path}.")
```

AA-Turner commented on Tools/scripts/verify_ensurepip_wheels.py:
```diff
@@ -0,0 +1,96 @@
+ENSURE_PIP_ROOT = Path(__file__).parent.parent.parent / "Lib/ensurepip"
+WHEEL_DIR = ENSURE_PIP_ROOT / "_bundled"
+ENSURE_PIP_INIT_PY_TEXT = (ENSURE_PIP_ROOT / "__init__.py").read_text(encoding="utf-8")
+GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS")
```
For explicitness?
```suggestion
GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS", False)
```

illia-v replied:
I believe checking its presence is enough. On GitHub Actions the value will be `"true"` as a string. I don't think it is worth changing it to `GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS") == "true"` either.

AA-Turner replied:
In that case, maybe `"GITHUB_ACTIONS" in os.environ` -- makes it clear that we don't care about the value at all.

A

illia-v replied:
On the second thought, I changed it to `GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS") == "true"` to be explicit.

AA-Turner replied:
Sorry, missed this comment -- I still think the simpler check is better.

A

AA-Turner commented on Misc/NEWS.d/next/Tests/2022-03-14-23-28-17.bpo-47016.K-t2QX.rst:
```diff
@@ -0,0 +1,2 @@
+Create a GitHub Actions workflow for verifying bundled pip and setuptools.
+Patch by Illia Volochii and Adam Turner.
```
Thanks!

AA-Turner reviewed (commented):
A few comments (I like the function refactoring!)

A

illia-v committed:
Refactor the script even more

illia-v committed:
Make `GITHUB_ACTIONS` a boolean

illia-v committed:
Stop using `actions/setup-python`

v4 will require specifying a Python version, system's Python should be enough for the task.

AA-Turner commented on .github/workflows/verify-ensurepip-wheels.yml:
```diff
@@ -17,6 +17,5 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - uses: actions/setup-python@v3
```
Why was this removed? We should bump to v4:

```yaml
- uses: actions/setup-python@v4
  with:
    python-version: '3'
```

illia-v replied:
I could either make the change you suggest or use Python from ubuntu-latest that is enough, I chose the latter.

illia-v replied:
https://github.com/python/cpython/runs/6938767354?check_suite_focus=true verifies that this workflow works without `actions/setup-python`.

AA-Turner replied:
I agree that it works, but it loses consistency with all the other workflows that use (rather than build) python, and ties us to the Python update cadence that Ubuntu use. I would argue that it is also more discoverable on the GHA UI, rather than implicitly using Ubuntu's default Python.

A

pradyunsg replied:
Besides, unlike Ubuntu's Python, setup-python's builds don't have esoteric patches that deviate behaviour in various ways.

illia-v replied:
Fair enough

AA-Turner commented on Tools/scripts/verify_ensurepip_wheels.py:
```diff
@@ -0,0 +1,98 @@
+ENSURE_PIP_ROOT = Path(__file__).parent.parent.parent / "Lib/ensurepip"
+WHEEL_DIR = ENSURE_PIP_ROOT / "_bundled"
+ENSURE_PIP_INIT_PY_TEXT = (ENSURE_PIP_ROOT / "__init__.py").read_text(encoding="utf-8")
+GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS") == "true"
```
A simpler check:
```suggestion
GITHUB_ACTIONS = "GITHUB_ACTIONS" in os.environ
```

illia-v committed:
Make changes to more files invoke the workflow

illia-v committed:
Make the workflow use `actions/setup-python` again

illia-v committed:
Merge branch 'main' into bpo-47016
